implemented disallow-variable-names

alawatthe opened the pull request:
disallowVariableNames: first implementation

disallowVariableNames takes an array of strings, which are considered as illegal variable names.

Fixes #41, #612, #918

alawatthe commented:
So here is a little longer description:

The issue of restricting the set of legal variable names has come up in the past several times, see #41, #612, #918.

The feedback in the discussions for such a rule was positive (I think), but no one created a pull request.


Therefore I'm starting with a pull request.

One open point in the past (see #612) was, if (and if "yes" how?) single letter variable names should be handled differently depending on whether they occur in loops or not.
In my opinion this problem can't be solved in a general way.
This is the reason why my pull request does not include any special handling for single letter variable names.


What do you think?

P.S.: This is my first rule I implemented from ground up, so don't spare me with critic :-)
For example I did know if there is a specified order for all the ```registerRule``` functions.

mrjoelkemp commented on test/rules/disallow-variable-names.js:
```diff
@@ -0,0 +1,18 @@
+        checker.registerDefaultRules();
+        checker.configure({ disallowVariableNames: ['temp', 'foo'] });
+    });
+    it('should report illegal variable names', function() {
```
spaces before every it please

mrjoelkemp commented on test/rules/disallow-variable-names.js:
```diff
@@ -0,0 +1,18 @@
+    });
+    it('should report illegal variable names', function() {
+        assert(checker.checkString('temp = 1;').getErrorCount() === 1);
+        assert(checker.checkString('obj[foo] = 1;').getErrorCount() === 1);
```
Can you also add tests for vars in loops and functions for completeness?

mrjoelkemp commented on test/rules/disallow-variable-names.js:
```diff
@@ -0,0 +1,18 @@
+        assert(checker.checkString('obj[foo] = 1;').getErrorCount() === 1);
+    });
+    it('should not report legal variable names', function() {
+        assert(checker.checkString('good = 1').isEmpty());
```
Same comment about completeness of the tests for other scenarios where vars are declared.

mrjoelkemp commented:
Do you need to rebase off of master?

mrjoelkemp commented on lib/config/configuration.js:
```diff
@@ -616,6 +616,8 @@ Configuration.prototype.registerDefaultRules = function() {
 
     this.registerRule(require('../rules/disallow-semicolons'));
+
+    this.registerRule(require('../rules/disallow-variable-names'));
```
No specific order. This is fine.

mrjoelkemp commented:
Nice work @alawatthe!

alawatthe commented:
@mrjoelkemp thank you, for your feedback :-)
I rebased, added the new lines and more tests. If you have more ideas for tests, I can add them, too.

mrjoelkemp commented on lib/rules/disallow-variable-names.js:
```diff
@@ -0,0 +1,57 @@
+    check: function(file, errors) {
+        var disallowVariableNames = this._disallowVariableNames;
+        file.iterateNodesByType('Identifier', function(node) {
+            if (disallowVariableNames.indexOf(node.name) !== -1) {
```
The iteration over disallowVariableNames is redundant. In `configure`, you could create an object of var name -> true (ex: `foo`: true) for a constant time lookup. You'd then do `if (disallowVariableNames[node.name])`.

mrjoelkemp replied:
What about uppercase var names? Your check should be case insensitive. I guess an alternative to the object lookup above is to create a regex from all of the disallowed vars and do a case insensitive match on node.name.

mrjoelkemp replied:
Be sure to add tests for the uppercase var names.

alawatthe replied:
My take on the uppercase/lowercase issue was, that one should be able to prohibit the use of ```foo``` but allow the use of ```Foo```. If you want to prohibit the use of ```Foo```, too, you can just add it to the array. Of course, this would still allow ```fOo``` and so on. And thus one could say this leads to exponential complexity...
That said, I can be persuaded on this one :-D
Are there any other opinions on this?

On the object lookup: Yes, the object lookup is probably faster. But this feels a little bit premature opimization like. I am the last person to not improve the code performance wise, but then we should do it right and do performance testing as well. Let me do some experiments with this.
But depending on the uppercase/lowercase issue I can of course implement a object lookup.

mrjoelkemp replied:
@mikesherov any opinion on the uppercase/lowercase issue above?

@alawatthe I was iffy on mentioning the object notation for the same fear of "the root of all evil," but:

1. This is one of the more obvious improvements that doesn't obfuscate the code (it actually clarifies the condition).
2. We have other rules that do this: https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/disallow-dangling-underscores.js#L84-L87, https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/disallow-implicit-type-conversion.js#L41-L44, https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/disallow-keywords.js#L31-L34, and a bunch more.

We'll wait for other opinions on casing to decide on the course of action.

mikesherov replied:
@alawatthe I agree variable should be case sensitive match. With regards to hash lookup instead of indexOf I agree with @mrjoelkemp that our idiomatic way is using hashes instead of array lookups. Please change.

alawatthe replied:
I changed it an object lookup.

Just for the record: I wasn't against the object lookup, I just think we can to do more performance wise. :-)

alawatthe commented:
Sorry for the noise on this bug :-/

I have been thinking a bit about the name for this rule. Since it also prevents property names like ```obj.bad``` a better name for the rule would be disallowIndentifierNames, i think.

I already changed the name and added tests on my local branch. So, if you agree with the name change I can upload the code in no time.

mikesherov commented:
@alawatthe sounds good.

mikesherov commented:
@alawatthe can you all rebase while you're at it? Thanks!

alawatthe commented:
I screwed the rebase :weary:. Therefore I created a completely new pull request https://github.com/jscs-dev/node-jscs/pull/960.

Sorry for the inconvenience :confused:
